include/engine/routing_algorithms/routing_base_mld.hpp

```diff
@@ -0,0 +1,283 @@
+#ifndef OSRM_ENGINE_ROUTING_BASE_MLD_HPP
+#define OSRM_ENGINE_ROUTING_BASE_MLD_HPP
+
+#include "engine/algorithm.hpp"
+#include "engine/datafacade/contiguous_internalmem_datafacade.hpp"
+#include "engine/routing_algorithms/routing_base.hpp"
+#include "engine/search_engine_data.hpp"
+
+#include "util/typedefs.hpp"
+
+#include <boost/assert.hpp>
+
+namespace osrm
+{
+namespace engine
+{
+namespace routing_algorithms
+{
+namespace mld
+{
+
+template <bool DIRECTION>
+void routingStep(const datafacade::ContiguousInternalMemoryDataFacade<algorithm::MLD> &facade,
+                 const partition::MultiLevelPartitionView &partition,
+                 const partition::CellStorageView &cells,
```

Comment:
We can add an closure here that makes it easy to compute the query level for a node:

```
struct QueryLevel
{
   const MultiLevelPartition &mlp;
   const PhantomNode &source;
   const PhantomNode &target;
   LevelID operator()(NodeID node)
  {
    auto forward_level = mlp.GetQueryLevel(source.foward_node, target.forward_node, node);
    auto backward_level = mlp.GetQueryLevel(source.backward_node, target.backward_node, node);
    return std::min(forward_level, backward_level);
  }
};
```

If we can fix out partition to ensure that the level 1 cell for `forward_node` and `backward_node` is always the same, we can optimize to only compute one.